src/dbnode/storage/shard.go

```diff
@@ -2650,6 +2658,78 @@ func (s *dbShard) Repair(
 	return repairer.Repair(ctx, nsCtx, nsMeta, tr, s)
 }
 
+func (s *dbShard) AggregateTiles(
+	ctx context.Context,
+	reader fs.DataFileSetReader,
+	sourceNsID ident.ID,
+	sourceBlockStart time.Time,
+	sourceShard databaseShard,
+	opts AggregateTilesOptions,
+	wOpts series.WriteOptions,
+) (int64, error) {
+	latestSourceVolume, err := sourceShard.latestVolume(sourceBlockStart)
+	if err != nil {
+		return 0, err
+	}
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   sourceNsID,
+			Shard:       sourceShard.ID(),
+			BlockStart:  sourceBlockStart,
+			VolumeIndex: latestSourceVolume,
+		},
+		FileSetType: persist.FileSetFlushType,
+		//TODO add after streaming supported - OrderByIndex: true
+	}
+	if err := reader.Open(openOpts); err != nil {
+		return 0, err
+	}
+	defer reader.Close()
+
+	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
+	bytesReader := bytes.NewReader(nil)
+	dataPointIter := m3tsz.NewReaderIterator(bytesReader, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
+	var lastWriteError error
+	var processedBlockCount int64
+
+	for {
+		id, tags, data, _, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return processedBlockCount, err
```

Comment:
Should we have partial results here or just `return 0, err`?

Reply:
I think partial was intended here. Any arguments against it?

Reply:
Because we do not rollback persisted data on some error, so partial result reflects the fact that some data has been made durable.

Reply:
Ah yeah I see.. how would retries work? Just rewrite all points and let cold flush upsert figure it out?

Reply:
Cold flush is not used in the latest implementation. By default, a retry would overwrite everything with a new volume. But I guess we will need something more sophisticated, like marking the shards that have been aggregated successfully, in order to avoid repetitive overwrite.